File: example/src/main/java/com/stripe/example/activity/PaymentSessionActivity.java
package com.stripe.example.activity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.stripe.android.CustomerSession;
import com.stripe.android.PaymentSession;
import com.stripe.android.PaymentSessionConfig;
import com.stripe.android.PaymentSessionData;
import com.stripe.android.StripeError;
import com.stripe.android.model.Address;
import com.stripe.android.model.Customer;
import com.stripe.android.model.PaymentMethod;
import com.stripe.android.model.ShippingInformation;
import com.stripe.android.model.ShippingMethod;
import com.stripe.android.view.ShippingInfoWidget;
import com.stripe.example.R;
import com.stripe.example.controller.ErrorDialogHandler;
import com.stripe.example.service.ExampleEphemeralKeyProvider;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Currency;
import java.util.Locale;

import static com.stripe.android.PayWithGoogleUtils.getPriceString;
import static com.stripe.android.view.PaymentFlowExtras.EVENT_SHIPPING_INFO_PROCESSED;
import static com.stripe.android.view.PaymentFlowExtras.EVENT_SHIPPING_INFO_SUBMITTED;
import static com.stripe.android.view.PaymentFlowExtras.EXTRA_DEFAULT_SHIPPING_METHOD;
import static com.stripe.android.view.PaymentFlowExtras.EXTRA_IS_SHIPPING_INFO_VALID;
import static com.stripe.android.view.PaymentFlowExtras.EXTRA_SHIPPING_INFO_DATA;
import static com.stripe.android.view.PaymentFlowExtras.EXTRA_VALID_SHIPPING_METHODS;

/**
 * An example activity that handles working with a {@link PaymentSession}, allowing you to collect
 * information needed to request payment for the current customer.
 */
public class PaymentSessionActivity extends AppCompatActivity {

    private BroadcastReceiver mBroadcastReceiver;
    private ErrorDialogHandler mErrorDialogHandler;
    private PaymentSession mPaymentSession;
    private ProgressBar mProgressBar;
    private TextView mResultTextView;
    private TextView mResultTitleTextView;
    private Button mSelectPaymentButton;
    private Button mSelectShippingButton;
    private PaymentSessionData mPaymentSessionData;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_payment_session);
        mProgressBar = findViewById(R.id.customer_progress_bar);
        mProgressBar.setVisibility(View.VISIBLE);
        mSelectPaymentButton = findViewById(R.id.btn_select_payment_method_aps);
        mSelectShippingButton = findViewById(R.id.btn_start_payment_flow);
        mErrorDialogHandler = new ErrorDialogHandler(this);
        mResultTitleTextView = findViewById(R.id.tv_payment_session_data_title);
        mResultTextView = findViewById(R.id.tv_payment_session_data);

        // CustomerSession only needs to be initialized once per app.
        final CustomerSession customerSession = setupCustomerSession();
        setupPaymentSession(customerSession);

        final LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.getInstance(this);
        mBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(@NonNull Context context, @NonNull Intent intent) {
                final ShippingInformation shippingInformation = intent
                        .getParcelableExtra(EXTRA_SHIPPING_INFO_DATA);
                final Intent shippingInfoProcessedIntent =
                        new Intent(EVENT_SHIPPING_INFO_PROCESSED);
                if (!isValidShippingInfo(shippingInformation)) {
                    shippingInfoProcessedIntent.putExtra(EXTRA_IS_SHIPPING_INFO_VALID, false);
                } else {
                    final ArrayList<ShippingMethod> shippingMethods = createSampleShippingMethods();
                    shippingInfoProcessedIntent.putExtra(EXTRA_IS_SHIPPING_INFO_VALID, true);
                    shippingInfoProcessedIntent.putParcelableArrayListExtra(
                            EXTRA_VALID_SHIPPING_METHODS, shippingMethods);
                    shippingInfoProcessedIntent.putExtra(EXTRA_DEFAULT_SHIPPING_METHOD,
                            shippingMethods.get(1));
                }
                localBroadcastManager.sendBroadcast(shippingInfoProcessedIntent);
            }

            private boolean isValidShippingInfo(@NonNull ShippingInformation shippingInfo) {
                return shippingInfo.getAddress() != null &&
                        Locale.US.getCountry().equals(shippingInfo.getAddress().getCountry());
            }
        };
        localBroadcastManager.registerReceiver(mBroadcastReceiver,
                new IntentFilter(EVENT_SHIPPING_INFO_SUBMITTED));
        mSelectPaymentButton.setOnClickListener(v ->
                mPaymentSession.presentPaymentMethodSelection(true));
        mSelectShippingButton.setOnClickListener(v -> mPaymentSession.presentShippingFlow());
    }

    @NonNull
    private CustomerSession setupCustomerSession() {
        CustomerSession.initCustomerSession(this,
                new ExampleEphemeralKeyProvider(new ProgressListenerImpl(this)));
        final CustomerSession customerSession = CustomerSession.getInstance();
        customerSession.retrieveCurrentCustomer(
                new InitialCustomerRetrievalListener(this));
        return customerSession;
    }

    private void setupPaymentSession(@NonNull CustomerSession customerSession) {
        mPaymentSession = new PaymentSession(this);
        final boolean paymentSessionInitialized = mPaymentSession.init(
                new PaymentSessionListenerImpl(this, customerSession),
                new PaymentSessionConfig.Builder()
                        .setPrepopulatedShippingInfo(getExampleShippingInfo())
                        .setHiddenShippingInfoFields(ShippingInfoWidget.PHONE_FIELD,
                                ShippingInfoWidget.CITY_FIELD)
                        .build());
        if (paymentSessionInitialized) {
            mPaymentSession.setCartTotal(2000L);
        }
    }

    @NonNull
    private String formatStringResults(@NonNull PaymentSessionData data) {
        final Currency currency = Currency.getInstance("USD");
        final StringBuilder stringBuilder = new StringBuilder();

        if (data.getPaymentMethod() != null) {
            final PaymentMethod paymentMethod = data.getPaymentMethod();
            final PaymentMethod.Card card = paymentMethod.card;

            if (card != null) {
                stringBuilder.append("Payment Info:\n").append(card.brand)
                        .append(" ending in ")
                        .append(card.last4)
                        .append(data.isPaymentReadyToCharge() ? " IS " : " IS NOT ")
                        .append("ready to charge.\n\n");
            }
        }
        if (data.getShippingInformation() != null) {
            stringBuilder.append("Shipping Info: \n");
            stringBuilder.append(data.getShippingInformation());
            stringBuilder.append("\n\n");
        }
        if (data.getShippingMethod() != null) {
            stringBuilder.append("Shipping Method: \n");
            stringBuilder.append(data.getShippingMethod()).append('\n');
            if (data.getShippingTotal() > 0) {
                stringBuilder.append("Shipping total: ")
                        .append(getPriceString(data.getShippingTotal(), currency));
            }
        }

        return stringBuilder.toString();
    }

    @NonNull
    private ArrayList<ShippingMethod> createSampleShippingMethods() {
        final ArrayList<ShippingMethod> shippingMethods = new ArrayList<>();
        shippingMethods.add(new ShippingMethod("UPS Ground", "ups-ground",
                "Arrives in 3-5 days", 0, "USD"));
        shippingMethods.add(new ShippingMethod("FedEx", "fedex",
                "Arrives tomorrow", 599, "USD"));
        return shippingMethods;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mPaymentSession.handlePaymentData(requestCode, resultCode, data);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mPaymentSession.onDestroy();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mBroadcastReceiver);
    }

    @NonNull
    private ShippingInformation getExampleShippingInfo() {
        final Address address = new Address.Builder()
                .setCity("San Francisco")
                .setCountry("US")
                .setLine1("123 Market St")
                .setLine2("#345")
                .setPostalCode("94107")
                .setState("CA")
                .build();
        return new ShippingInformation(address, "Fake Name", "(555) 555-5555");
    }

    private void onPaymentSessionDataChanged(@NonNull CustomerSession customerSession,
                                             @NonNull PaymentSessionData data) {
        mPaymentSessionData = data;
        mProgressBar.setVisibility(View.VISIBLE);
        customerSession.retrieveCurrentCustomer(
                new PaymentSessionChangeCustomerRetrievalListener(this));
    }

    private static final class PaymentSessionListenerImpl
            extends PaymentSession.ActivityPaymentSessionListener<PaymentSessionActivity> {
        @NonNull private final CustomerSession mCustomerSession;

        private PaymentSessionListenerImpl(@NonNull PaymentSessionActivity activity,
                                           @NonNull CustomerSession customerSession) {
            super(activity);
            mCustomerSession = customerSession;
        }

        @Override
        public void onCommunicatingStateChanged(boolean isCommunicating) {
            final PaymentSessionActivity activity = getListenerActivity();
            if (activity == null) {
                return;
            }

            activity.mProgressBar
                    .setVisibility(isCommunicating ? View.VISIBLE : View.INVISIBLE);
        }

        @Override
        public void onError(int errorCode, @NonNull String errorMessage) {
            final PaymentSessionActivity activity = getListenerActivity();
            if (activity == null) {
                return;
            }

            activity.mErrorDialogHandler.show(errorMessage);
        }

        @Override
        public void onPaymentSessionDataChanged(@NonNull PaymentSessionData data) {
            final PaymentSessionActivity activity = getListenerActivity();
            if (activity == null) {
                return;
            }

            activity.onPaymentSessionDataChanged(mCustomerSession, data);
        }
    }

    private static final class InitialCustomerRetrievalListener
            extends CustomerSession.ActivityCustomerRetrievalListener<PaymentSessionActivity> {
        private InitialCustomerRetrievalListener(@NonNull PaymentSessionActivity activity) {
            super(activity);
        }

        @Override
        public void onCustomerRetrieved(@NonNull Customer customer) {
            final PaymentSessionActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            activity.mProgressBar.setVisibility(View.INVISIBLE);
        }

        @Override
        public void onError(int httpCode, @NonNull String errorMessage,
                            @Nullable StripeError stripeError) {
            final PaymentSessionActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            activity.mErrorDialogHandler.show(errorMessage);
            activity.mProgressBar.setVisibility(View.INVISIBLE);
        }
    }

    private static final class PaymentSessionChangeCustomerRetrievalListener
            extends CustomerSession.ActivityCustomerRetrievalListener<PaymentSessionActivity> {
        private PaymentSessionChangeCustomerRetrievalListener(
                @NonNull PaymentSessionActivity activity) {
            super(activity);
        }

        @Override
        public void onCustomerRetrieved(@NonNull Customer customer) {
            final PaymentSessionActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            activity.mProgressBar.setVisibility(View.INVISIBLE);
            activity.mSelectPaymentButton.setEnabled(true);
            activity.mSelectShippingButton.setEnabled(true);

            if (activity.mPaymentSessionData != null) {
                activity.mResultTitleTextView.setVisibility(View.VISIBLE);
                activity.mResultTextView.setText(
                        activity.formatStringResults(activity.mPaymentSessionData));
            }
        }

        @Override
        public void onError(int httpCode, @NonNull String errorMessage,
                            @Nullable StripeError stripeError) {
            final PaymentSessionActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            activity.mProgressBar.setVisibility(View.INVISIBLE);
        }
    }

    private static final class ProgressListenerImpl
            implements ExampleEphemeralKeyProvider.ProgressListener {
        @NonNull private final WeakReference<PaymentSessionActivity> mActivityRef;

        private ProgressListenerImpl(@NonNull PaymentSessionActivity activity) {
            mActivityRef = new WeakReference<>(activity);
        }

        @Override
        public void onStringResponse(@NonNull String response) {
            final PaymentSessionActivity activity = mActivityRef.get();
            if (activity != null && response.startsWith("Error: ")) {
                activity.mErrorDialogHandler.show(response);
            }
        }
    }
}

File: example/src/main/java/com/stripe/example/activity/PaymentSessionActivity.kt
package com.stripe.example.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.vectordrawable.graphics.drawable.VectorDrawableCompat
import com.stripe.android.CustomerSession
import com.stripe.android.PaymentSession
import com.stripe.android.PaymentSessionConfig
import com.stripe.android.PaymentSessionData
import com.stripe.android.core.StripeError
import com.stripe.android.model.Address
import com.stripe.android.model.Customer
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.ShippingInformation
import com.stripe.android.model.ShippingMethod
import com.stripe.android.view.BillingAddressFields
import com.stripe.android.view.PaymentUtils
import com.stripe.example.R
import com.stripe.example.databinding.PaymentSessionActivityBinding
import java.util.Currency
import java.util.Locale

/**
 * An example activity that handles working with a [PaymentSession], allowing you to collect
 * information needed to request payment for the current customer.
 */
class PaymentSessionActivity : AppCompatActivity() {
    private val viewBinding: PaymentSessionActivityBinding by lazy {
        PaymentSessionActivityBinding.inflate(layoutInflater)
    }

    private val viewModel: ActivityViewModel by viewModels()

    private val customerSession: CustomerSession by lazy {
        CustomerSession.getInstance()
    }

    private lateinit var paymentSession: PaymentSession
    private val notSelectedText: String by lazy {
        getString(R.string.not_selected)
    }
    private val snackbarController: SnackbarController by lazy {
        SnackbarController(viewBinding.coordinator)
    }

    private var paymentSessionData: PaymentSessionData? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(viewBinding.root)

        paymentSession = createPaymentSession(savedInstanceState == null)

        viewBinding.selectPaymentMethodButton.setOnClickListener {
            paymentSession.presentPaymentMethodSelection()
        }
        viewBinding.startPaymentFlowButton.setOnClickListener {
            paymentSession.presentShippingFlow()
        }

        viewModel.paymentSessionDataResult.observe(
            this,
            { result ->
                result.fold(
                    onSuccess = {
                        onPaymentSessionDataChanged(it)
                    },
                    onFailure = {
                        snackbarController.show(it.message.orEmpty())
                    }
                )
            }
        )
        viewModel.customerResult.observe(
            this,
            { result ->
                result.fold(
                    onSuccess = {
                        onCustomerRetrieved()
                    },
                    onFailure = {
                        hideProgressBar()
                    }
                )
            }
        )

        viewModel.isProcessing.observe(
            this,
            { isCommunicating ->
                when (isCommunicating) {
                    true -> disableUi()
                    false -> enableUi()
                }
            }
        )
    }

    private fun createPaymentSession(
        shouldPrefetchCustomer: Boolean = false
    ): PaymentSession {
        if (shouldPrefetchCustomer) {
            disableUi()
        } else {
            enableUi()
        }

        return PaymentSession(
            activity = this,
            config = PaymentSessionConfig.Builder()
                .setPaymentMethodsFooter(R.layout.add_payment_method_footer)
                .setAddPaymentMethodFooter(R.layout.add_payment_method_footer)
                .setPrepopulatedShippingInfo(EXAMPLE_SHIPPING_INFO)
                .setHiddenShippingInfoFields()
                // Optionally specify the `PaymentMethod.Type` values to use.
                // Defaults to `PaymentMethod.Type.Card`
                .setPaymentMethodTypes(listOf(PaymentMethod.Type.Card))
                .setShouldShowGooglePay(true)
                .setAllowedShippingCountryCodes(setOf("US", "CA"))
                .setShippingInformationValidator(ShippingInformationValidator())
                .setShippingMethodsFactory(ShippingMethodsFactory())
                .setWindowFlags(WindowManager.LayoutParams.FLAG_SECURE)
                .setBillingAddressFields(BillingAddressFields.PostalCode)
                .setShouldPrefetchCustomer(shouldPrefetchCustomer)
                .setCanDeletePaymentMethods(true)
                .build()
        ).also {
            it.init(listener = viewModel.paymentSessionListener)
            it.setCartTotal(2000L)
        }
    }

    private fun createPaymentMethodDescription(data: PaymentSessionData): String {
        val paymentMethod = data.paymentMethod
        return when {
            paymentMethod != null -> {
                paymentMethod.card?.let { card ->
                    "${card.brand} ending in ${card.last4}"
                } ?: paymentMethod.type?.code.orEmpty()
            }
            data.useGooglePay -> {
                "Use Google Pay"
            }
            else -> {
                notSelectedText
            }
        }
    }

    private fun createShippingInfoDescription(shippingInformation: ShippingInformation?): String {
        return if (shippingInformation != null) {
            listOfNotNull(
                shippingInformation.name,
                shippingInformation.address?.line1,
                shippingInformation.address?.line2,
                shippingInformation.address?.city,
                shippingInformation.address?.state,
                shippingInformation.address?.country,
                shippingInformation.address?.postalCode,
                shippingInformation.phone
            ).joinToString("\n")
        } else {
            notSelectedText
        }
    }

    private fun createShippingMethodDescription(shippingMethod: ShippingMethod?): String {
        return if (shippingMethod != null) {
            listOfNotNull(
                shippingMethod.label,
                shippingMethod.detail,
                PaymentUtils.formatPriceStringUsingFree(
                    shippingMethod.amount,
                    Currency.getInstance("USD"),
                    "Free"
                )
            ).joinToString("\n")
        } else {
            notSelectedText
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        paymentSession.handlePaymentData(requestCode, resultCode, data ?: Intent())
    }

    private fun onPaymentSessionDataChanged(
        data: PaymentSessionData
    ) {
        paymentSessionData = data
        disableUi()
        customerSession.retrieveCurrentCustomer(viewModel.customerSessionListener)
    }

    private fun enableUi() {
        hideProgressBar()
        viewBinding.selectPaymentMethodButton.isEnabled = true
        viewBinding.startPaymentFlowButton.isEnabled = true
    }

    private fun disableUi() {
        viewBinding.progressBar.visibility = View.VISIBLE
        viewBinding.selectPaymentMethodButton.isEnabled = false
        viewBinding.startPaymentFlowButton.isEnabled = false
    }

    private fun onCustomerRetrieved() {
        enableUi()

        paymentSessionData?.let { paymentSessionData ->
            viewBinding.readyToCharge.setCompoundDrawablesRelativeWithIntrinsicBounds(
                VectorDrawableCompat.create(
                    resources,
                    when (paymentSessionData.isPaymentReadyToCharge) {
                        true -> R.drawable.ic_check
                        false -> R.drawable.ic_cancel
                    },
                    null
                ),
                null,
                null,
                null
            )

            viewBinding.paymentMethod.text =
                createPaymentMethodDescription(paymentSessionData)

            viewBinding.shippingInfo.text =
                createShippingInfoDescription(paymentSessionData.shippingInformation)

            viewBinding.shippingMethod.text =
                createShippingMethodDescription(paymentSessionData.shippingMethod)
        }
    }

    private class ShippingInformationValidator : PaymentSessionConfig.ShippingInformationValidator {
        override fun isValid(shippingInformation: ShippingInformation): Boolean {
            return setOf(Locale.US.country, Locale.CANADA.country)
                .contains(shippingInformation.address?.country.orEmpty())
        }

        override fun getErrorMessage(shippingInformation: ShippingInformation): String {
            return "The country must be US or Canada."
        }
    }

    private class ShippingMethodsFactory : PaymentSessionConfig.ShippingMethodsFactory {
        override fun create(
            shippingInformation: ShippingInformation
        ): List<ShippingMethod> {
            return when (shippingInformation.address?.country) {
                "CA" -> SHIPPING_METHODS_CA
                else -> SHIPPING_METHODS_US
            }
        }
    }

    private fun hideProgressBar() {
        viewBinding.progressBar.visibility = View.INVISIBLE
    }

    internal class ActivityViewModel : ViewModel() {
        val isProcessing = MutableLiveData<Boolean>()
        val paymentSessionDataResult = MutableLiveData<Result<PaymentSessionData>>()

        val customerResult = MutableLiveData<Result<Customer>>()

        val customerSessionListener: CustomerSession.CustomerRetrievalListener by lazy {
            object : CustomerSession.CustomerRetrievalListener {
                init {
                    BackgroundTaskTracker.onStart()
                }

                override fun onCustomerRetrieved(customer: Customer) {
                    BackgroundTaskTracker.onStop()
                    customerResult.value = Result.success(customer)
                }

                override fun onError(
                    errorCode: Int,
                    errorMessage: String,
                    stripeError: StripeError?
                ) {
                    BackgroundTaskTracker.onStop()
                    customerResult.value = Result.failure(RuntimeException())
                }
            }
        }

        val paymentSessionListener: PaymentSession.PaymentSessionListener by lazy {
            object : PaymentSession.PaymentSessionListener {
                override fun onCommunicatingStateChanged(isCommunicating: Boolean) {
                    if (isCommunicating) {
                        BackgroundTaskTracker.onStart()
                    }

                    isProcessing.value = isCommunicating
                }

                override fun onError(errorCode: Int, errorMessage: String) {
                    BackgroundTaskTracker.onStop()
                    paymentSessionDataResult.value = Result.failure(RuntimeException(errorMessage))
                }

                override fun onPaymentSessionDataChanged(data: PaymentSessionData) {
                    BackgroundTaskTracker.onStop()
                    paymentSessionDataResult.value = Result.success(data)
                }
            }
        }
    }

    private companion object {
        private val EXAMPLE_SHIPPING_INFO = ShippingInformation(
            Address.Builder()
                .setCity("San Francisco")
                .setCountry("US")
                .setLine1("123 Market St")
                .setLine2("#345")
                .setPostalCode("94107")
                .setState("CA")
                .build(),
            "Jenny Rosen",
            "(555) 555-5555"
        )

        private val SHIPPING_METHODS_CA = listOf(
            ShippingMethod(
                "Canada Post",
                "canada-post",
                599,
                "CAD",
                "Arrives in 3-5 days"
            )
        )

        private val SHIPPING_METHODS_US = listOf(
            ShippingMethod(
                "UPS Ground",
                "ups-ground",
                599,
                "USD",
                "Arrives in 3-5 days"
            ),
            ShippingMethod(
                "FedEx Overnight",
                "fedex",
                1499,
                "USD",
                "Arrives tomorrow"
            )
        )
    }
}
